arbnode/transaction_streamer.go

```diff
@@ -754,6 +785,36 @@ func (s *TransactionStreamer) countDuplicateMessages(
 					}
 					dbMessageParsed.Message.BatchGasCost = batchGasCostBkup
 				}
+				msg := nextMessage.MessageWithMeta.Message
+				if arbos.IsEspressoMsg(dbMessageParsed.Message) &&
+					arbos.IsEspressoMsg(msg) &&
+					!bytes.Equal(msg.L2msg, dbMessageParsed.Message.L2msg) {
+					// Check to see if the difference is the existence of block merkle proof in the new message,
+					// The batcher can append this on the fly (see AddEspressoBlockMerkleProof).
+					// If this is the case, update the database with the block merkle justification.
+					_, newJst, err := arbos.ParseEspressoMsg(msg)
+					if err != nil {
+						return 0, false, nil, err
+
+					}
+					_, oldJst, err := arbos.ParseEspressoMsg(dbMessageParsed.Message)
+					if err != nil {
+						return 0, false, nil, err
+					}
+					if oldJst.BlockMerkleJustification == nil && newJst.BlockMerkleJustification != nil {
```

Comment:
@ImJeremyHe @nomaxg why do we need to have this extra logic to check for duplicates? We should add a comment to the code.

Reply:
Agree that there should be more comments.

That was because that batch poster will wait for the justification from hotshot and will fill the messages with it. This means that actually the batch poster are modifying the existing messages.

For some validators, they will receive the l2 messages without any jst first and maintain the l2 blocks. And they will also fetch the messages from L1. No matter where the messages are from, they will eventually be stored in the `transaction streamer`. So, the transaction streamer will check if the these messages from different sources are the same. If not, it will consider a reorg happening.

So we have to add these code here to get around the misunderstanding of reorg